src/Collections/PriorityQueue.cs

```diff
@@ -0,0 +1,100 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Bearded.Utilities.Collections
+{
+    /// <summary>
+    /// A priority queue that is implemented with a binary min-heap.
+    /// </summary>
+    /// <typeparam name="TPriority"></typeparam>
+    /// <typeparam name="TValue"></typeparam>
+    public sealed class PriorityQueue<TPriority, TValue> : StaticPriorityQueue<TPriority, TValue> where TPriority : IComparable<TPriority>
+    {
+        private readonly Dictionary<TValue, int> valueDict = new Dictionary<TValue, int>();
+
+        /// <summary>
+        /// Creates a new instance of a priority queue.
+        /// </summary>
+        public PriorityQueue() { }
+
+        /// <summary>
+        /// Creates a new instance of a priority queue with the specified initial capacity.
+        /// </summary>
+        /// <param name="capacity">Initial capacity of the priority queue.</param>
+        public PriorityQueue(int capacity)
+            : base(capacity) { }
+
+        /// <summary>
+        /// Creates a new instance of a priority queue with the specified initial data.
+        /// </summary>
+        /// <param name="data">Initial data to fill the priority queue with.</param>
+        public PriorityQueue(IEnumerable<KeyValuePair<TPriority, TValue>> data)
+        {
+            this.data = data.ToArray();
+            this.Count = this.data.Length;
+            for (int i = 0; i < this.data.Length; i++)
+                this.valueDict.Add(this.data[i].Value, i);
+            for (int i = this.data.Length / 2 - 1; i >= 0; i--)
+                this.cascadeDown(i);
+        }
+
+        /// <summary>
+        /// Decreases the priority of the specified element [O(log n)].
+        /// </summary>
+        /// <param name="value">The element that should change.</param>
+        /// <param name="newPriority">The new priority of the element.</param>
+        public void DecreasePriority(TValue value, TPriority newPriority)
+        {
+            int i = this.valueDict[value];
+            if (this.data[i].Key.CompareTo(newPriority) == -1)
```

Comment:
This should be replaced by a `< 0` check.
CompareTo is not required to return -1/1. The documentation specifies only that the returned value has to be negative or positive.